polkit/validate/validate.go

```diff
@@ -0,0 +1,297 @@
+	ExtraChildren []xml.Name `xml:",any"`
+}
+
+type policyConfig struct {
+	XMLName xml.Name
+	Element
+
+	Vendor    []Element `xml:"vendor"`
+	VendorURL []Element `xml:"vendor_url"`
+	IconName  []Element `xml:"icon_name"`
+
+	Actions []action `xml:"action"`
+}
+
+type action struct {
+	Element
+
+	ID string `xml:"id,attr"`
+
+	Vendor    []Element `xml:"vendor"`
+	VendorURL []Element `xml:"vendor_url"`
+	IconName  []Element `xml:"icon_name"`
+
+	Description []message  `xml:"description"`
+	Message     []message  `xml:"message"`
+	Defaults    []defaults `xml:"defaults"`
+	Annotate    []annotate `xml:"annotate"`
+}
+
+type message struct {
+	Element
+
+	GettextDomain string `xml:"gettext-domain,attr"`
+	Language      string `xml:"lang,attr"` // to match xml:lang
+}
+
+type defaults struct {
+	Element
+
+	AllowAny      []Element `xml:"allow_any"`
+	AllowInactive []Element `xml:"allow_inactive"`
+	AllowActive   []Element `xml:"allow_active"`
+}
+
+type annotate struct {
+	Element
+
+	Key string `xml:"key,attr"`
+}
+
+func ValidatePolicy(r io.Reader) (actionIDs []string, err error) {
+	decoder := xml.NewDecoder(r)
+	var config policyConfig
+	if err := decoder.Decode(&config); err != nil {
+		return nil, err
+	}
+	// check for additional data after the root element
+	switch err := decoder.Decode(new(interface{})); err {
+	case nil:
+		return nil, fmt.Errorf("invalid XML: additional data after root element")
+	case io.EOF:
+		// nothing
+	default:
+		return nil, err
+	}
+
+	return validateConfig(config)
+}
+
+func validateConfig(config policyConfig) ([]string, error) {
+	if config.XMLName != (xml.Name{Local: "policyconfig"}) {
+		return nil, fmt.Errorf("root element must be <policyconfig>")
+	}
+
+	if err := validateNoExtraElements(config.Element, "<policyconfig>"); err != nil {
+		return nil, err
+	}
+	if err := validateNoCharData(config.Element, "<policyconfig>"); err != nil {
+		return nil, err
+	}
+
+	if err := validateOptionalProperty(config.Vendor, "<vendor>", "<policyconfig>"); err != nil {
+		return nil, err
+	}
+	if err := validateOptionalProperty(config.VendorURL, "<vendor_url>", "<policyconfig>"); err != nil {
+		return nil, err
+	}
+	if err := validateOptionalProperty(config.IconName, "<icon_name>", "<policyconfig>"); err != nil {
+		return nil, err
+	}
+
+	seenIDs := make(map[string]struct{})
+	for _, a := range config.Actions {
+		if err := validateAction(a, seenIDs); err != nil {
+			return nil, err
+		}
+	}
+
+	actionIDs := make([]string, 0, len(seenIDs))
+	for id := range seenIDs {
+		actionIDs = append(actionIDs, id)
+	}
+	return actionIDs, nil
+}
+
+func validateNoExtraElements(element Element, name string) error {
+	if len(element.ExtraAttrs) != 0 {
+		return fmt.Errorf("%s element contains unexpected attributes", name)
+	}
+	if len(element.ExtraChildren) != 0 {
+		return fmt.Errorf("%s element contains unexpected children", name)
+	}
+	return nil
+}
+
+func validateNoCharData(element Element, name string) error {
```

Comment:
I would merge these two helpers into one that takes some form of flags about which checks to do, the main reason I'm asking for this is that afaict the second one is always used after the first, also it compacts the important but not so interesting portion of each validation function here

Reply:
I've combined these two into a single function, taking a flags argument that can be set to allowCharData.